ministryofjustice/correspondence_tool_staff
# CT-2581 Add git-crypt to Track a Query codebase

## Description
First step in making this application deployable on Cloud Platform - store sensitive credentials locally

## Notes
- Introduces git-crypt with 2 collaborators (@mseedat-moj, @Polsonby)
- Introduces gem `Overcommit` to manage git hooks, specifically pre-commit hook
- Introduces dummy secrets for eventual Kubernetes required secrets
- Changes to readme
 
## Self-review checklist

 
### Screenshots
![Screen Shot 2019-11-20 at 13 41 55](https://user-images.githubusercontent.com/47113046/69245472-c66d1f80-0b9e-11ea-84e5-48112c39c4e0.png)

 
### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-2581
 
### Deployment
Will require `git-crypt` to be installed on the user system plus the following steps

IMPORTANT
All developers need to run bundle install, then initialise the overcommit utility to add git hooks. Otherwise there is no protection from committing unencrypted secrets. There are three steps
1. Run `bundle install`
```
ian ~/dev/moj/cts CT-2581-add-git-crypt $ bundle
The dependency tzinfo-data (>= 0) will be unused by any of the platforms Bundler is installing for. Bundler is installing for ruby but the dependency is only for x86-mingw32, x86-mswin32, x64-mingw32, java. To add those platforms to the bundle, run `bundle lock --add-platform x86-mingw32 x86-mswin32 x64-mingw32 java`.
Fetching gem metadata from https://rubygems.org/.........
Using rake 12.3.3
...snip...
Bundle complete! 94 Gemfile dependencies, 233 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
Post-install message from overcommit:
Install hooks by running `overcommit --install` in your Git repository
ian ~/dev/moj/cts CT-2581-add-git-crypt $
```

2. Run `bundle exec overcommit --install` 
```
ian ~/dev/moj/cts CT-2581-add-git-crypt $ be overcommit --install
Installing hooks into /Users/iananderson/dev/moj/cts
Successfully installed hooks into /Users/iananderson/dev/moj/cts
ian ~/dev/moj/cts CT-2581-add-git-crypt $
```

3. Tell overcommit to trust the new hook that's been installed by running `overcommit --sign pre-commit`
```
iananderson@MOJL0419LM ~/dev/moj/cts CT-2581-add-git-crypt $ bundle exec overcommit --sign pre-commit
Updated signature of plugin CustomScript

iananderson@MOJL0419LM ~/dev/moj/cts CT-2581-add-git-crypt $ git commit -m 'message'
Running pre-commit hooks
Run CustomScript.......................................[CustomScript] OK

✓ All pre-commit hooks passed

Running commit-msg hooks
Check subject capitalization.....................[CapitalizedSubject] WARNING
Subject should start with a capital letter
Check for trailing periods in subject................[TrailingPeriod] OK
Check subject line................................[SingleLineSubject] OK
Check text width..........................................[TextWidth] OK

⚠ All commit-msg hooks passed, but with warnings

[CT-2581-add-git-crypt 92dcc6fb] woot3
 1 file changed, 1 insertion(+)
 rename config/kubernetes/production/{supersecret.yml => supersecrrret.yml} (67%)
iananderson@MOJL0419LM ~/dev/moj/cts CT-2581-add-git-crypt $ 

```

### Manual testing instructions
- On checkout of this file, check `config/kubernetes/demo/secrets.yaml` it should be encrypted gibberish and require `git-crypt unlock` to see the hidden message
- Add a file anywhere in the directory called `fake-secrets.lol` and then attempt to commit this file. An error should prevent this file from being committed
